subworkflows/local/deduplicate.nf

```diff
@@ -0,0 +1,170 @@
+                meta, bam, bai ->
+                    meta2 = [:]
+                    meta2.reference = meta.reference
+                [ meta2, meta, bam, bai ]
+            }
+            .combine(
+                by:0,
+                ch_refs
+            )
+            .multiMap{
+                ignore_me, meta, bam, bai, meta2, fasta, fasta_fai ->
+                bam: [ meta, bam ]
+                fasta: fasta
+                fasta_fai: fasta_fai
+            }
+
+            // Dedup each bam
+            PICARD_MARKDUPLICATES(
+                ch_markduplicates_input.bam,
+                ch_markduplicates_input.fasta,
+                ch_markduplicates_input.fasta_fai
+            )
+            ch_versions             = ch_versions.mix( PICARD_MARKDUPLICATES.out.versions.first() )
+
+            ch_dedupped_region_bam  = PICARD_MARKDUPLICATES.out.bam
+
+    } else if ( params.deduplication_tool == "dedup" ) {
+        ch_dedup_input = BAM_SPLIT_BY_REGION.out.bam_bai
+            .map {
+                meta, bam, bai ->
+                [ meta, bam ]
+            }
+
+        DEDUP( ch_dedup_input )
+        ch_versions            = ch_versions.mix( DEDUP.out.versions.first() )
+
+        ch_dedupped_region_bam = DEDUP.out.bam
+    }
+
+ch_input_for_samtools_merge = ch_dedupped_region_bam
+    .map {
+        meta, bam ->
+        meta2 = meta.clone().findAll{ it.key != 'genomic_region' }
+        [ meta2, bam ]
+    }
+    .groupTuple()
+    .map {
+        meta, bam ->
+            ref_meta = [:]
+            ref_meta.reference = meta.reference
+        [ ref_meta, meta, bam ]
+    }
+    .combine(
+        by:0,
+        ch_refs
+    )
+    .multiMap{
+        // bam here is a list of bams
+        ignore_me, meta, bam, meta2, fasta, fasta_fai ->
+        bam:        [ meta, bam ]
+        fasta:      fasta
+        fasta_fai:  fasta_fai
+    }
+
+    // Merge the bams for each region into one bam
+    SAMTOOLS_MERGE(
+        ch_input_for_samtools_merge.bam,
+        ch_input_for_samtools_merge.fasta,
+        ch_input_for_samtools_merge.fasta_fai
+    )
+    ch_versions   = ch_versions.mix( SAMTOOLS_MERGE.out.versions )
+
+
+    // Sort the merged bam and index
+    SAMTOOLS_SORT ( SAMTOOLS_MERGE.out.bam )
+    ch_versions   = ch_versions.mix( SAMTOOLS_SORT.out.versions )
+    ch_dedup_bam  = SAMTOOLS_SORT.out.bam
+
+    SAMTOOLS_INDEX ( ch_dedup_bam )
+    ch_versions   = ch_versions.mix( SAMTOOLS_INDEX.out.versions )
+    ch_dedup_bai  =  params.fasta_largeref ? SAMTOOLS_INDEX.out.csi : SAMTOOLS_INDEX.out.bai
+
+    // Finally run flagstat on the dedupped bam
+    ch_input_for_samtools_flagstat = ch_dedup_bam.join( ch_dedup_bai )
+
+    SAMTOOLS_FLAGSTAT(
+        ch_input_for_samtools_flagstat
+    )
+    ch_versions       = ch_versions.mix( SAMTOOLS_FLAGSTAT.out.versions )
+
+    ch_dedup_flagstat = SAMTOOLS_FLAGSTAT.out.flagstat
```

Comment:
Why not mix with `ch_multiqc_files` and export that (this is the common pattern across subworkflwos/workflwos)

Reply:
The flagstat files need to go to `endorspy` eventually, no?

Reply:
Yes good point